src/Pagination/Paginator.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace App\Pagination;
+
+use Doctrine\ORM\QueryBuilder as DoctrineQueryBuilder;
+use Doctrine\ORM\Tools\Pagination\CountWalker;
+use Doctrine\ORM\Tools\Pagination\Paginator as DoctrinePaginator;
+
+/**
+ * @author Javier Eguiluz <javier.eguiluz@gmail.com>
+ */
+class Paginator
+{
+    private const PAGE_SIZE = 10;
+    private $currentPage;
+    private $pageSize;
+    private $results;
+    private $numResults;
+
+    public function __construct(DoctrineQueryBuilder $queryBuilder, int $currentPage = 1, int $pageSize = self::PAGE_SIZE)
+    {
+        $this->currentPage = max(1, $currentPage);
+        $this->pageSize = $pageSize;
+        $firstResult = ($this->currentPage - 1) * $pageSize;
+
+        $query = $queryBuilder
+            ->setFirstResult($firstResult)
+            ->setMaxResults($pageSize)
+            ->getQuery();
+
+        if (0 === \count($queryBuilder->getDQLPart('join'))) {
+            $query->setHint(CountWalker::HINT_DISTINCT, false);
+        }
+
+        $paginator = new DoctrinePaginator($query, true);
```

Comment:
this second argument should be configurable. It is quite frequent to have paginated queries which don't fetch join a collection. and the pagination can be more efficient in that case (using a single query rather than 2)

Reply:
btw, if there is no join at all in the query (detected above), we are 100% sure that there is no fetch join of a collection (not all joins are collection fetch joins, so it is still worth it to allow configuring it for other cases)

Reply:
I'd prefer to not make this configurable. The difference between making 1 or 2 queries should not be noticeable, mostly because the extra query will be very simple.

If not doing this introduces some bugs or problems, then I'm willing to make it configurable. Thanks.